### ISSUE TITLE
Add AspNet Core side shim for System.Web.VirtualPathUtility

### ISSUE THREAD

**Issue by CZEMacLeod** [dotnet/systemweb-adapters#83]
## Summary

Add the `System.Web.VirtualPathUtility` class to systemweb adapters to allow the use from AspNet Core and class libraries whilst migrating an application.

## Motivation and goals

Within our application we make use of this class, mainly to handle dynamic paths and in areas where we need to return absolute paths regardless of the request url (e.g. in webapi call, middleware, partial views etc.).

## In scope

Handle the following methods on `System.Web.VirtualPathUtility`
```cs
namespace System.Web;

public static class VirtualPathUtility
{
  public static string Combine(string basePath, string relativePath);
  public static string GetDirectory(string virtualPath);
  public static string GetExtension(string virtualPath);
  public static string GetFileName(string virtualPath);
  public static bool IsAppRelative(string virtualPath);
  public static string MakeRelative(string fromPath, string toPath);
  public static string ToAbsolute(string virtualPath);
  public static string ToAppRelative(string virtualPath);
}
```

## Out of scope

The overloads which allow you to include the `applicationPath` would not be needed (at least not for my project).

## Risks / unknowns

Most of these functions relate to translating paths starting ~/ to the base url which should be available with #81 as HttpRuntime.AppDomainAppVirtualPath.
The naïve implementation could use UrlBuilder and/or combinations of StartsWith and SubString to implement the main functions.
The Get* methods would probably need to get the IFileProvider interface registered to get the appropriate physical file paths and names for virtual paths.
GetFileName and GetExtension should work regardless of whether the file resolved to a physical file or not.
Most of these functions are no longer in use in web applications since it is now common to have each web application at the root, rather than installed as a sub site in IIS so this is largely for legacy projects where these abstractions have not yet been removed, or where configuration files or similar use 'app relative paths' like ~/App_Data/myconfig.json

## Examples

```cs
public static class ThemeClaims
{
	public static string GetThemeOrDefault(this ClaimsIdentity identity, string @default = "default")
	{
		var claim = identity?.FindFirst("myapp:Theme");
		if (claim is null) return @default;
		return claim.Value;
	}
}

public class ThemeModel
{
	public string Theme { get; set; } = 
		(System.Web.HttpContext.Current.User?.Identity as ClaimsIdentity).GetThemeOrDefault();
	public string ThemeURL { 
		get => System.Web.VirtualPathUtility
			.ToAbsolute($"~/custom/themes/{Theme}/theme.css"); 
	}
}
```

**Comment by CZEMacLeod** [comment]
I think a lot of the implementation for this is already in `System.Web.Util.UrlPath`. It is just that that is a) not public and b) not in AspNet Core.

**Comment by adityamandaleeka** [comment]
Triage: This sounds reasonable. @CZEMacLeod would you be interested in contributing some/all of the implementation for this?

**Comment by CZEMacLeod** [comment]
@adityamandaleeka I've had a look at this and I could certainly put together something if it was based on the code / logic in the net48 `System.Web.Util.UrlPath` - but I don't believe that is under any form of open-source agreement. I guess the mono project would have a version, but again I'm not sure about the copyright for that.
I could certainly look at (re)implementing this from scratch, based solely on the published docs, although I think it might easily miss some edge cases covered in other implementations, especially as I am not aware of any sort of test suite/data to check the implementation against.

**Comment by CZEMacLeod** [comment]
@adityamandaleeka Some code here https://github.com/CZEMacLeod/System.Web.VirtualPathUtility - note that this is based partially on the net48 framework `System.Web.Util.UrlPath` implementation and I'm not entirely sure what the legalities are around that code.

**Comment by adityamandaleeka** [comment]
@CZEMacLeod, thanks for being cautious with licensing. The good news is that there's no issue from the licensing perspective to use code from the .NET Framework sources since they're under an acceptable license ([MIT](https://github.com/microsoft/referencesource/blob/master/LICENSE.txt)).

We might have to add a third-party notice entry in this repo. I'll follow up internally to take care of that.

**Comment by CZEMacLeod** [comment]
@adityamandaleeka Excellent.
Well there is an initial implementation of `VirtualPathUtility` for AspNet Core in the repo I mentioned previously. I started putting together some test cases, but haven't had time to go through all permutations or functions. There is a test library (netstandard2.0) with some test calls of `ToAbsolute`, and it is consumed by an AspNet Core net6.0 project and an AspNet/System.Web net48 project. *Same output on both.
I'm not quite sure what to do about a test plan, but I could probably create a PR with the code from the  [VirtualPathUtility.Library](https://github.com/CZEMacLeod/System.Web.VirtualPathUtility/tree/main/VirtualPathUtility.Library) project as is? 
Let me know how you want to proceed.

*It should be noted that the net48 implementation doesn't actually throw the error listed in the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.web.virtualpathutility.toabsolute?view=netframework-4.8). The framework throws an `ArgumentException` with a custom message, not an `ArgumentOutOfRangeException` for ToAbsolute with an invalid relative path. I'm not sure which is actually better, but it is a mismatch. I can probably change it to more closely match the net framework implementation if need be.

**Comment by adityamandaleeka** [comment]
@CZEMacLeod Feel free to open a PR for this and we'll review the approach. It may take a bit longer than usual now since the team is fairly occupied at the moment, but we'll get to it as soon as we can.

**Comment by twsouthwick** [comment]
Closed via #179
